File: py_opencv/splitmerge.py
import cv2 as cv
import numpy as np

img = cv.imread('photos/dog1.jpg')
cv.imshow('Dog', img)

blank = np.zeros(img.shape[:2], dtype='uint8')

b, g, r = cv.split(img)
cv.imshow('blue', b)
cv.imshow('green', g)
cv.imshow('red', r)

print(img.shape)
print(b.shape)
print(g.shape)
print(r.shape)

merged = cv.merge([b, g, r])
cv.imshow('merge', merged)
blue = cv.merge([b, blank, blank])
green = cv.merge([blank, g, blank])
red = cv.merge([blank, blank, r])

cv.imshow('bl', blue)
cv.imshow('gr', green)
cv.imshow('re', red)


cv.waitKey(0)

File: cpp_opencv/splitmerge/splitmerge.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <stdio.h>
#include <string>
#include <vector>

using namespace cv;

std::string photo = "../../photos/dog1.jpg";

int main(){
    Mat image = imread(photo);
    resize(image, image, cv::Size(), 0.2, 0.2);
    imshow("dog", image);

    Mat blank = Mat::zeros(Size(image.cols, image.rows), CV_8UC1);

    Mat bgrchannel[3]; 
    split(image, bgrchannel);
    imshow("b", bgrchannel[0]);
    imshow("g", bgrchannel[1]);
    imshow("r", bgrchannel[2]);

    std::vector<Mat> input_merge = {bgrchannel[0], bgrchannel[1], bgrchannel[2]};
    Mat merged; 
    merge(input_merge, merged);
    imshow("merged", merged);

    std::vector<Mat> bl = {bgrchannel[0], blank, blank};
    std::vector<Mat> gr = {blank, bgrchannel[1], blank};
    std::vector<Mat> re = {blank, blank, bgrchannel[2]};

    Mat blue, green, red;
    merge(bl, blue);
    merge(gr, green);
    merge(re, red);

    imshow("blue", blue);
    imshow("green", green);
    imshow("red", red);


    waitKey(0);
    return 0;
}

